src/node_crypto.h

```diff
@@ -136,18 +132,17 @@ class SecureContext : public BaseObject {
         cert_(nullptr),
         issuer_(nullptr) {
     MakeWeak<SecureContext>(this);
-    env->isolate()->AdjustAmountOfExternalAllocatedMemory(kExternalSize);
   }
 
   void FreeCTXMem() {
     if (ctx_) {
-      env()->isolate()->AdjustAmountOfExternalAllocatedMemory(-kExternalSize);
-      if (ctx_->cert_store == root_cert_store) {
+      if (SSL_CTX_get_cert_store(ctx_) == root_cert_store) {
         // SSL_CTX_free() will attempt to free the cert_store as well.
         // Since we want our root_cert_store to stay around forever
         // we just clear the field. Hopefully OpenSSL will not modify this
         // struct in future versions.
-        ctx_->cert_store = nullptr;
+        // XXX: Maybe we should just call X509_STORE_up_ref() instead?
+        SSL_CTX_set_cert_store(ctx_, nullptr);
```

Comment:
(I happened to be looking at this line today and wondered how this change handled the problem.)

I don't think this works: `SSL_CTX_set_cert_store` [will free](https://github.com/openssl/openssl/blob/207a9cb3522882d1e9dc764c921425ba47a6def6/ssl/ssl_lib.c#L3550) the existing `X509_STORE` before assigning `nullptr`. But the point of this code was to avoid that because `root_cert_store` is a global that shouldn't be freed.

Rather I think the answer is to use `X509_STORE_up_ref` before assigning `root_cert_store` in the first place. Sadly, in 1.0.1, `X509_STORE` contained a reference count but it was ignored. In 1.0.2 it's used, but there's no good way to _increment_ it. It's only in 1.1.0 that `X509_STORE_up_ref` was finally added. (Albeit with an unhelpful signtaure that doesn't return the pointer itself.)

(I'm also quite suspicious of the ownership semantics of `root_cert_store` anyway.)

I expect I'll send a change this week to try and clear this up and will try to #ifdef things so that it'll work with 1.1.0 too.